src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs

```diff
@@ -575,5 +575,42 @@ private static IEnumerable<PropertyInfo> GetAllProperties(Type type)
 
             return allProperties;
         }
+
+        private static object GetPropertyValue(PropertyInfo property, object instance, IConfiguration config, BinderOptions options)
+        {
+            object propertyValue = property.GetValue(instance);
+            string propertyName = GetPropertyAttributeKeyName(property);
+
+            if (string.IsNullOrEmpty(propertyName))
+            {
+                propertyName = property.Name;
+            }
+
+            return BindInstance(property.PropertyType, propertyValue, config.GetSection(propertyName), options);
+        }
+
+        private static string GetPropertyAttributeKeyName(MemberInfo property)
+        {
+            if (property == null)
+            {
+                return string.Empty;
+            }
+
+            foreach (var attributeData in property.GetCustomAttributesData())
+            {
+                if (attributeData.AttributeType != typeof(ConfigurationKeyNameAttribute))
+                {
+                    continue;
+                }
+
+                var constructorArgument = attributeData.ConstructorArguments.FirstOrDefault();
```

Comment:
Should this throw if the attribute gets out of sync with this code, since the attribute is a ConfigurationKeyName at this point, so we shouldn't have fallback behavior to empty string, something really bad has happened if this gets out of sync right?

Reply:
`if (constructorArgument.ArgumentType == typeof(string))` if this is ever not true it should throw rather than fall through

Reply:
Updated the flow around attribute checking. I actually created the GetPropertyName to provide a shim for future changes where other bindings could be necessary. I did refactor this functionality to take a default value if GetPropertyValue should fail. Very similar to the other GetValue functions around it.